plugins/bf4/bf4.cpp

```diff
@@ -0,0 +1,242 @@
+            peekProc((BYTE *) pModule + 0x21B80C0, host) && // Host value: "IP:Port" when in a server, "bot" when loading and empty when it's hidden.
+            peekProc((BYTE *) pModule + 0x24AFAE5, team) && // Team value: US (United States); RU (Russia); CH (China).
+            peekProc((BYTE *) squad_offset_2 + 0x230, squad) && // Squad value: 0 (not in a squad); 1 (Alpha); 2 (Bravo); 3 (Charlie)... 26 (Zulu).
+            peekProc((BYTE *) squad_offset_2 + 0x234, squad_leader); // Team value: US (United States); RU (Russia); CH (China).
+
+    // This prevents the plugin from linking to the game in case something goes wrong during values retrieval from memory addresses.
+    if (! ok) {
+        return false;
+    }
+
+    if (state) { // If not in-game
+        context.clear(); // Clear context
+        identity.clear(); // Clear identity
+        // Set vectors values to 0.
+        for (int i=0;i<3;i++) {
+            avatar_pos[i] = avatar_front[i] = avatar_top[i] = camera_pos[i] =  camera_front[i] = camera_top[i] = 0.0f;
+        }
+
+        return true; // This tells Mumble to ignore all vectors.
+    }
+
+    serverid[sizeof(serverid)-1] = 0; // NUL terminate queried C strings. We do this to ensure the strings from the game are NUL terminated. They should be already, but we can't take any chances.
+    std::string ServerID(serverid);
+    ServerID = jsonStringEscape(ServerID); // Replace '"' character with ' ' (space) to prevent JSON break.
+    std::ostringstream ocontext;
+    ocontext << " {\"Server ID\": \"" << ServerID << "\"}"; // Set context with server ID
+    context = ocontext.str();
+
+    std::wostringstream oidentity;
+    oidentity << "{";
+    host[sizeof(host)-1] = 0; // NUL terminate queried C strings. We do this to ensure the strings from the game are NUL terminated. They should be already, but we can't take any chances.
+    std::string Host(host);
+    Host = jsonStringEscape(Host);
+    if (Host.find("bot") == std::string::npos) // Set host string as empty if "bot" is found in it.
+        if (!Host.empty()) {
+            oidentity << std::endl << "\"Host\": \"" << host << "\""; // If it's not empty, set Host (IP:Port) in identity.
+        }
+
+    team[sizeof(team)-1] = 0; // NUL terminate queried C strings. We do this to ensure the strings from the game are NUL terminated. They should be already, but we can't take any chances.
+    std::string Team(team);
+    Team = jsonStringEscape(Team);
+    if (!Team.empty()) {
+        oidentity << std::endl;
+        if (Team == "US")
+            oidentity << "\"Team\": \"United States\""; // If team value is US, set "United States" as team in identity.
+        else if (Team == "CH")
+            oidentity << "\"Team\": \"China\""; // If team value is CH, set "China" as team in identity.
+        else if (Team == "RU")
+            oidentity << "\"Team\": \"Russia\""; // If team value is RU, set "Russia" as team in identity.
+    }
+
+    if (squad == 0) {
```

Comment:
I think this code should be restructured.

Remove this check here and unindent the lines below it.

So the code becomes:

	if (squad == 1) {
		oidentity << std::endl << "\"Squad\": \"Alpha\",";
	} else if (squad == 2)
		oidentity << std::endl << "\"Squad\": \"Bravo\",";
	[...]
	} else {
		oidentity << std::endl << "\"Squad\": null,"
	}

This makes the code handle invalid values (values that are 0, or > 26) consistent, and simplifies the code.